fix(ci): disk validation for docker volume mount

gustavovalverde opened the pull request:
## Reviewer Instructions

Since this change is urgent, please put suggested cleanups in ticket #7677.

## Motivation

PR #7662 did not work as expected, as the validation was too simplistic.

@teor2345 also found that Google recently updated their `cos-stable` image and this upgraded Docker considerably (by 4 major versions), which might be the reason of this appearing recently with no changes to the code. 

Fixes: #7659

## Solution

- Revert to use `busybox` instead of `ubuntu`, as this was causing errors in the VM logs
- Before formatting, validate the disk is not being used with `lsof`
- Do not mount the Docker volume before running Docker. Instead mount it in the `docker run` command
  -  Based on new Docker versions, this approach can now be used: https://docs.docker.com/storage/volumes/#how-mounting-block-devices-works

## Review

- We did not run this multiple times in the previous PR, we should re-run the tests multiple times to confirm.
- Since this change is urgent, please put suggested cleanups in ticket #7677

### Reviewer Checklist

  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

gustavovalverde committed:
fix(ci): disk validation for docker volume mount

teor2345 commented:
@gustavovalverde since this is an urgent fix, do you mind if I work on it today?
It would be helpful to get it merged before the full sync jobs start tomorrow, so we can test all these bug fixes.

teor2345 committed:
Use a symlink for lightwalletd cached state rather than mounting the same volume twice

teor2345 committed:
Avoid "sdb seems to be busy" errors from docker by adding extra sleeps

teor2345 committed:
Add a missing backslash

teor2345 commented:
## Reviewer Instructions

Since this change is urgent, please put suggested cleanups in ticket #7677

teor2345 committed:
Remove symlink from workflow

teor2345 committed:
Symlink lightwalletd path in entrypoint.sh

gustavovalverde commented:
@teor2345 we can also output the Docker daemon logs after the failure, to confirm if it gives more information about the error:

```shell
sudo journalctl -u docker.service

```

teor2345 committed:
Retry on failure and check Docker logs

teor2345 commented:
> @teor2345 we can also output the Docker daemon logs after the failure, to confirm if it gives more information about the error:
> 
> ```shell
> sudo journalctl -u docker.service
> ```

Thanks, I gave this a try.

teor2345 committed:
End ssh shell lines with explicit terminators

teor2345 committed:
Delete Docker containers if Docker mount fails

teor2345 committed:
Revert symlink changes in entrypoint.sh

teor2345 commented:
I'm currently retrying the commands if they fail, I also need to delete the containers that were created without the correct mounts.

teor2345 committed:
Debug using lsof

teor2345 commented:
I'm trying lsof for debugging. If that doesn't work, fuser, losetup, or lsof | grep a_inode might help:
https://stackoverflow.com/questions/624154/linux-which-process-is-causing-device-busy-when-doing-umount/45782183#45782183

This could be caused by an automatic fsck:
https://unix.stackexchange.com/questions/179855/device-already-mounted-or-resource-is-busy/342464#342464

teor2345 committed:
Use correct lsof commands

teor2345 committed:
Use correct syntax for lsof +D

teor2345 commented:
So far lsof isn't showing much: there aren't any files open on the device. Maybe one of the other commands will help more.

gustavovalverde committed:
fix(ci): make multiple validations before mounting

Loop and checks for three conditions:
The device `/dev/sdb` exists.
No process is using the device `/dev/sdb`.
No process is using the Docker volume directory.

gustavovalverde committed:
Merge branch 'main' into fix-busy-disk-mount

gustavovalverde committed:
fix: do not pre-mount docker volume

The Docker version available with the newer `cos-stable` OS (https://cloud.google.com/release-notes#cos-109-17800-0-45) allows to mount the image when running it.

Mounting it before makes the disk unavailable.

gustavovalverde committed:
fix: remove extra `;`

gustavovalverde committed:
fix: just confirm with `lsof` and show it's output

gustavovalverde commented:
I canceled [fix: remove extra ;] (https://github.com/ZcashFoundation/zebra/pull/7665/commits/3c29dab5c349acf3ac44dea4ce5c446ba91daef3) with a new commit

But it did not fail to mount the volumes

gustavovalverde committed:
chore: reduce diff
